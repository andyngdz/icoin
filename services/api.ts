import { Format } from 'services'
import { IAPI, IAssetResponse } from 'types'
import { PER_PAGE } from 'data'
import axios from 'axios'

const CoingeckoAPI = axios.create({
  baseURL: process.env.NEXT_PUBLIC_COINGECKO_API
})

const CoinCapAPI = axios.create({
  baseURL: process.env.NEXT_PUBLIC_COINCAP_API
})

const API: IAPI = {
  getGlobals: () => CoingeckoAPI.get('/global'),

  getAssets: async (
    params = {
      limit: PER_PAGE
    }
  ) => {
    const response = await CoinCapAPI.get<IAssetResponse>('/assets', {
      params
    })
    const { data } = response.data

    response.data.data = Format.toNumber(data, [
      'marketCapUsd',
      'priceUsd',
      'volumeUsd24Hr',
      'changePercent24Hr',
      'supply',
      'maxSupply',
      'vwap24Hr'
    ])

    return response
  }
}

export { API }
