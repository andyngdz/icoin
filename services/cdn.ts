const CDN = {
  getIcon: (id: string) =>
    `https://cdn.jsdelivr.net/gh/atomiclabs/cryptocurrency-icons@9ab8d6934b83a4aa8ae5e8711609a70ca0ab1b2b/svg/color/${id}.svg`,

  defaultIcon: `https://cdn.jsdelivr.net/gh/atomiclabs/cryptocurrency-icons@9ab8d6934b83a4aa8ae5e8711609a70ca0ab1b2b/svg/color/generic.svg`,
}

export { CDN }
