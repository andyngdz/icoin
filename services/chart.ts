/**
 * Author: Andy D. Ng <itc.anhduy@gmail.com>
 * Chart.js contains the useful function to manage a Chart
 */

import { ChartData, ChartOptions, TimeScale } from 'chart.js'
import { IAssetHistory, ICalculateInterval, TTime } from 'types'
import { merge } from 'lodash'
import { sub } from 'date-fns'
import { TIME_TO_INTERVAL } from 'data'
import ChartJS from 'chart.js'

/**
 * @description
 * We are using Roboto font and
 * Set it here for the ChartJS as well
 */
ChartJS.defaults.global.defaultFontFamily = "'Inter', sans-serif"

const Chart = {
  /**
   * @description Calculate interval post data by id, time, start and end date
   *
   * @param id The id of coin such as: bitcoin, ethereum
   *
   * @param time The time from the TimeSelection component - (Check TTime in chart.d.ts)
   *
   * @returns `ICalculateInterval`
   *
   * @example
   * ```
   *    const { time, onTimeChange } = useTime()
   *     const [getIntervals, { data }] = useLazyQuery<IAssetHistories>(COIN_CHART)
   *
   *     useEffect(() => {
   *       const interval = Chart.calculateInterval(id, time)
   *       getIntervals({ variables: interval })
   *     }, [time])
   *```
   */
  calculateInterval: (id: string, time: TTime): ICalculateInterval => {
    let subtractDate: Date
    const end = Date.now()

    switch (time) {
      case '1D':
        subtractDate = sub(end, { days: 1 })
        break
      case '1W':
        subtractDate = sub(end, { weeks: 1 })
        break
      case '1M':
        subtractDate = sub(end, { months: 1 })
        break
      case '3M':
        subtractDate = sub(end, { months: 3 })
        break
      case '6M':
        subtractDate = sub(end, { months: 6 })
        break
      case '1Y':
      case 'ALL':
        subtractDate = sub(end, { years: 1 })
        break
    }

    return {
      end,
      id,
      interval: TIME_TO_INTERVAL[time],
      start: subtractDate.valueOf()
    }
  },

  /**
   * @description Create time scale with unit for Line Chart
   *
   * @param time The time from the TimeSelection component - (Check TTime in chart.d.ts)
   *
   * @returns `TimeScale`
   */
  createTimeScale: (time: TTime): TimeScale => {
    const timeScale: TimeScale = {}

    switch (time) {
      case '1D':
        timeScale.unit = 'hour'
        break
      case '1W':
        timeScale.unit = 'day'
        break
      case '1M':
      case '3M':
        timeScale.unit = 'week'
        break
      case '6M':
      case '1Y':
      case 'ALL':
        timeScale.unit = 'month'
        break
    }

    return timeScale
  },

  /**
   * @description Create a new chart
   *
   * @param ctx The canvas rendering context
   *
   * @param assetHistories Contains the list history data such as: timestamp, priceUsd, etc...
   *
   * @param options The chart options, it will be merged with the default options
   *
   * @returns `ChartJS`
   */
  createNewChart: (
    canvas: HTMLCanvasElement,
    time: TTime,
    chartData: ChartData,
    options?: ChartOptions
  ): ChartJS => {
    const ctx = canvas.getContext('2d')

    return new ChartJS(ctx, {
      type: 'line',
      data: chartData,
      options: merge(
        {
          tooltips: {
            position: 'nearest'
          },
          legend: {
            display: false
          },
          scales: {
            xAxes: [
              {
                type: 'time',
                time: Chart.createTimeScale(time),
                gridLines: {
                  display: false
                }
              }
            ],
            yAxes: [
              {
                position: 'right',
                gridLines: {
                  drawBorder: false
                }
              }
            ]
          }
        },
        options
      )
    })
  },

  /**
   * @description Create chart data
   *
   * @param assetHistories Contains the list history data such as: timestamp, priceUsd, etc...
   *
   * @returns `ChartData`
   */
  createChartData: (assetHistories: Array<IAssetHistory>): ChartData => {
    const data = assetHistories.map(({ timestamp, priceUsd }) => ({
      t: timestamp,
      y: priceUsd
    }))

    const chartData: ChartData = {
      datasets: [
        {
          data
        }
      ]
    }

    return chartData
  }
}

export { Chart }
