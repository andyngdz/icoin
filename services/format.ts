import { formatCurrency } from '@coingecko/cryptoformat'
import numeral from 'numeral'

const numberFormatter = new Intl.NumberFormat('en-US')

const percentFormatter = new Intl.NumberFormat('en-US', {
  style: 'percent',
  minimumFractionDigits: 2,
  maximumFractionDigits: 2,
})

const Format = {
  bigNumber: (value: number) => numeral(value).format('($0.00a)'),

  currency: (value: number) => formatCurrency(value, 'USD', 'en'),

  percent: (value: number) => percentFormatter.format(value / 100),
}

export { Format, numberFormatter, percentFormatter }
