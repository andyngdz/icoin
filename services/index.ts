export * from './paginate'
export * from './chart'
export * from './format'
export * from './api'
export * from './cdn'
export * from './test-utils'
export * from './price-direction'
export * from './price'
export * from './routes'
