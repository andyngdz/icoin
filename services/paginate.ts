const Paginate = {
  offset: (page: number, perPage: number): number => --page * perPage,

  count: (total: number, perPage: number): number => Math.ceil(total / perPage)
}

export { Paginate }
