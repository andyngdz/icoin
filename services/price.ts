import { IAssetHistory } from 'types'
import { maxBy, minBy } from 'lodash'

const Price = {
  priceIterate: (history: IAssetHistory): number => history.priceUsd,

  max: (histories: IAssetHistory[]): number =>
    maxBy(histories, Price.priceIterate).priceUsd,

  min: (histories: IAssetHistory[]): number =>
    minBy(histories, Price.priceIterate).priceUsd
}

export { Price }
