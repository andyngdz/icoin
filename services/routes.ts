const Routes = {
  home: '/',
  exchanges: '/exchanges',
  trending: '/trending'
}

export { Routes }
