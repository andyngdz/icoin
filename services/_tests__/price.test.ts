import { histories } from 'mock-data'
import { Price } from 'services'

describe('Price', () => {
  const { data } = histories

  it('Should return max price', () => {
    expect(Price.max(data.assetHistories)).toMatchSnapshot()
  })

  it('Should return min price', () => {
    expect(Price.min(data.assetHistories)).toMatchSnapshot()
  })
})
