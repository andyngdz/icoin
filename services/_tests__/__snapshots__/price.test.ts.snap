// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Price Should return max price 1`] = `33782.07577413444`;

exports[`Price Should return min price 1`] = `33459.08978145529`;
